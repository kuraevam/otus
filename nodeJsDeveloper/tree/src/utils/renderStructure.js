function render(structure, depth = 0) {
  const rows = [];

  let str = structure.name;

  if (depth > 0) {
    str = `|${
      [...Array(depth - 1)].map((value, index) => {
        if (index === 0) {
          return ' '.repeat(3);
        }
        return `|${' '.repeat(2)}`;
      }).join('')
    }${depth > 1 ? '|' : '-'
    }${'-'.repeat(5)
    }${structure.name}`;
  }

  rows.push(str);

  if (Array.isArray(structure.items)) {
    depth += 1;
    for (const structureChild of structure.items) {
      const rowsChild = render(structureChild, depth);
      rows.push(...rowsChild);
    }
  }

  return rows;
}

export default function renderStructure(structure) {
  const rows = render(structure);
  return rows.join('\n');
}
