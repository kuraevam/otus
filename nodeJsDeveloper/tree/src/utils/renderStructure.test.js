import renderStructure from './renderStructure.js';

describe('renderStructure', () => {
  test('it should return the tree structure', async () => {
    const data = {
      name: 1,
      items: [{ name: 2, items: [{ name: 3 }, { name: 4 }] }, {
        name: 5,
        items: [{ name: 6, items: [{ name: 7, items: [{ name: 8, items: [{ name: 9 }] }] }] }],
      }],
    };

    const result = renderStructure(data);

    expect(result).toMatchSnapshot();
  });
});
