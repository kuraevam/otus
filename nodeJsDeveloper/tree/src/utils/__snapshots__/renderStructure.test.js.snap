// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderStructure it should return the tree structure 1`] = `
"1
|------2
|   |-----3
|   |-----4
|------5
|   |-----6
|   |  |-----7
|   |  |  |-----8
|   |  |  |  |-----9"
`;
